mongodb-inconsistent-indexes-across-shards.md (#101)
percona-platform/portal-doc: docs/advisors/checks/mongodb-inconsistent-indexes-across-shards.md

# MongoDB Sharding - Inconsistent Indexes Across Sharded Collections

## Description
This check warns if there are inconsistent indexes across shards for sharded collections. Missing or inconsistent indexes across the shards can have a negative impact on performance.

## Resolution

> A sharded collection has an inconsistent index if the collection does not have the exact same indexes (including the index options) on each shard that contains chunks for the collection. Although inconsistent indexes should not occur during normal operations, inconsistent indexes can occur in some situations. Examples of those situations can include the following:

Reviewer: ```suggestion
A sharded collection has an inconsistent index if the collection does not have the exact same indexes (including the index options) on each shard that contains chunks for the collection. Although inconsistent indexes should not occur during normal operations, they can occur in some situations. Examples of those situations are the following:
```

- When a user is creating an index with a unique key constraint and one shard contains a chunk with duplicate documents. In such cases, the create index operation may succeed on the shards without duplicates but fail on the shard with duplicates.

> - When a user is creating an index across the shards in a rolling manner (i.e. manually building the index individually across the shards one by one). The index could be inconsistent across all shards if the manual process either fails to build the index for an associated shard or incorrectly builds an index with different specifications for one or more shards. Indexes are often created this way when the shards are very large or when there are a high number of shards.

Reviewer: ```suggestion
- When a user is creating an index across the shards in a rolling manner - manually building the index individually across the shards one by one. The index could be inconsistent across all shards if the manual process either fails to build the index for an associated shard or incorrectly builds an index with different specifications for one or more shards. Indexes are often created this way when the shards are very large or when there are a high number of shards.
```

> Starting in MongoDB 4.2.6 and 4.4 and onwards, the config server primary checks for index inconsistencies across the shards for sharded collections by default.

Reviewer 2: ```suggestion
Starting with MongoDB 4.2.6 and 4.4, the config server primarily checks for index inconsistencies across the shards for sharded collections by default.
```

> Additionally running the serverStatus command on the config server primary will return the field shardedIndexConsistency to report on the number of sharded collections with index inconsistencies.

Reviewer 2: ```suggestion
Additionally, running the **serverStatus** command on the config server primary will return the field **shardedIndexConsistency** to report the number of sharded collections with index inconsistencies.
```

> If shardedIndexConsistency reports any index inconsistencies, you can identify the missing index or any inconsistent properties across the collection’s shards by running a specific  aggregation pipeline that returns results on inconsistencies. That pipeline script can be found in the [Manage Indexes section of MongoDB Documentation](https://www.mongodb.com/docs/manual/tutorial/manage-indexes/).

Reviewer 2: ```suggestion
If **shardedIndexConsistency** reports any index inconsistencies, you can identify the missing index or any inconsistent properties across the collection’s shards by running a specific aggregation pipeline that returns results on inconsistencies. You can find this pipeline script in the [Manage Indexes section of the MongoDB Documentation](https://www.mongodb.com/docs/manual/tutorial/manage-indexes/).
```

> If shardedIndexConsistency reports any index inconsistencies, you can identify the missing index or any inconsistent properties across the collection’s shards by running a specific  aggregation pipeline that returns results on inconsistencies. That pipeline script can be found in the “Manage Indexes” section here:  [MongoDB Documentation](https://www.mongodb.com/docs/manual/tutorial/manage-indexes/).

Reviewer: ```suggestion
If shardedIndexConsistency reports any index inconsistencies, you can identify the missing index or any inconsistent properties across the collection’s shards by running a specific  aggregation pipeline that returns results on inconsistencies. That pipeline script can be found in the [Manage Indexes section of MongoDB Documentation](https://www.mongodb.com/docs/manual/tutorial/manage-indexes/).
```

> **To resolve the inconsistency where an index is missing from the collection on a particular shard(s), you can either -** 

Reviewer: ```suggestion
## Resolution

**The inconsistency where an index is missing from the collection on a particular shard(s)** 

Perform either of the following steps: 
```

Author: @nastena1606 We need to keep a like this only, since there are  variants of index inconsistency and each has it's own resolution.
Also "##Resolution" is mentioned at the top of the documentation.

Reviewer: @Paragbhayani1 got your point. IMHO the description under the Resolution section distracts the user from the steps needed to be taken. Perhaps it makes sense to separate them into separate sections?

Reviewer: I would place it under description. This seems logical

Author: @nastena1606  ok, sure we can do that.

> - Perform a rolling index build for the collection on the affected shard(s) to make sure indexes are built consistently across all involved shards.

Reviewer 2: ```suggestion
- Perform a rolling index build for the collection on the affected shards to make sure indexes are built consistently across all involved shards.
```

(OR)

> - Issue an index build db.collection.createIndex() from a mongos instance. The operation only builds the collection's index on the shard(s) missing the index.

Reviewer 2: ```suggestion
- Issue an index build **db.collection.createIndex()** from a mongos instance. The operation only builds the collection's index on the shards missing the index.
```

> **To resolve where the index properties differ across the shards,**

Reviewer: ```suggestion
**The index properties differ across the shards,**
```

Author: @nastena1606 We need to keep a like this only, since there are variants of index inconsistency and each has it's own resolution.
Also "##Resolution" is mentioned at the top of the documentation.

> Drop the incorrect index from the collection on the affected shard(s) and rebuild the index by using the above method.

Reviewer 2: ```suggestion
Drop the incorrect index from the collection on the affected shards and rebuild the index by using the above method.
```

Reviewer 2: https://developers.google.com/style/plurals-parentheses

# MongoDB sharding - inconsistent indexes across sharded collections

## Description
This check warns if there are inconsistent indexes across shards for sharded collections. Missing or inconsistent indexes across the shards can have a negative impact on performance.

> ## Resolution

Reviewer: ```suggestion

```
